src/core/Akka.Tests/Actor/ReceiveActorHandlersTests.cs

```diff
@@ -0,0 +1,209 @@
+            handlers.AddGenericReceiveHandler<bool>(_ => true, _ => true));
+    }
+
+    // TODO Confirm use case - This is theoretically a breaking change. Conceptually it should not be because Object handler
+    // with no predicate is the same as a ReceiveAny handler.
+    [Fact]
+    public void Given_ObjectTypedReceiveHandlerWithNoPredicate_When_Adding_AnyReceiveHandler_Then_Should_Fail()
+    {
+        var handlers = new ReceiveActorHandlers();
+        handlers.AddTypedReceiveHandler(typeof(object), null, _ => true);
+
+        // This should throw because the object handler is already added and would catch this before.
+        Assert.Throws<InvalidOperationException>(() =>
+            handlers.AddReceiveAnyHandler(_ => { }));
+    }
+
+    [Fact]
+    public void Given_GenericReceiveHandlerWithPredicate_When_Adding_SameGenericReceiveHandlerWithPredicate_Then_Should_Succeed()
+    {
+        var handlers = new ReceiveActorHandlers();
+        handlers.AddGenericReceiveHandler<int>(_ => true, _ => true);
+
+        // The handler added has a predicate which makes it uncertain if it will handle the message.
+        // Adding another handler for the same type combination should be allowed.
+        handlers.AddGenericReceiveHandler<int>(null, _ => true);
+    }
+
+    [Fact]
+    public void Given_TypedReceiveHandler_When_Adding_DifferentTypedReceiveHandler_Then_Should_Succeed()
+    {
+        var handlers = new ReceiveActorHandlers();
+        handlers.AddTypedReceiveHandler(typeof(string), _ => true, _ => true);
+
+        handlers.AddTypedReceiveHandler(typeof(int), _ => true, _ => true);
+    }
+
+    [Fact]
+    public void Given_GenericReceiveHandler_When_Adding_DifferentGenericReceiveHandler_Then_Should_Succeed()
+    {
+        var handlers = new ReceiveActorHandlers();
+        handlers.AddGenericReceiveHandler<string>(null, _ => true);
+
+        handlers.AddGenericReceiveHandler<int>(_ => true, _ => true);
+    }
+
+    [Fact]
+    public void Given_TypedReceiveHandlerWithPredicate_When_Adding_DifferentTypedReceiveHandlerWithPredicate_Then_Should_Succeed()
+    {
+        var handlers = new ReceiveActorHandlers();
+        handlers.AddTypedReceiveHandler(typeof(object), _ => true, _ => true);
+
+        // This should be allowed because the object handler is already but it has a predicate that might not match.
+        handlers.AddTypedReceiveHandler(typeof(int), _ => true, _ => true);
+    }
+    
+    /*
+     * IFoo
+     * Bar: IFoo
+     *
+     * Receive<IFoo>
+     * Receive<Bar>
+     */
+    
+    private interface IFoo { }
+    private class Bar : IFoo { }
+    private class Baz : IFoo { }
+    
+    private static readonly Predicate<IFoo> FooPredicate = _ => true;
+    private static readonly Predicate<Baz> BazPredicate = _ => true;
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void Given_TypedReceiveHandler_can_match_interface_on_ConcreteTypes(bool usePredicate)
```

Comment:
This spec at the bottom covers two scenarios:

1. Tests the preservation of Akka.NET (,1.5] `Receive<T>` declaration order
2. Asserts that your `Type`-assignability code from https://github.com/akkadotnet/akka.net/pull/7498 works